### PR TITLE
librbd: prevent concurrent AIO callbacks to external clients

### PR DESCRIPTION
**Pull request by dillaman**
With the various threads and conditional IO paths within librbd, it was
possible for external AIO callbacks to be concurrently executed from
different librbd threads. These callbacks should be serialized to reduce
the unexpected potential for data corruption.

Fixes: http://tracker.ceph.com/issues/40417
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/librbd/io/AioCompletion.cc [hunk 1]
```diff
@@ -242,5 +246,34 @@ ssize_t AioCompletion::get_return_value() {
...
+    ictx->external_callback_in_progress.store(false);
+    if (queue_empty) {
+      // we have looped at least twice on an empty queue which should avoid
+      // a possible race between the pop loop and resetting the in-progress
+      // state
+      break;
+    }
+    queue_empty = true;
+  }
+}
```
[the remark this answers is not in the document]
Comment (dillaman) [reply src/librbd/io/AioCompletion.cc]: Yeah, that should work.

**Comment by dillaman** [comment]
> we could modify `RBD::AioCompletion` constructor to wrap `complete_cb`

I would have generally preferred something like that as well, but it would have required allocating a second object on the heap during `RBD::AioCompletion` contruction to store the original `complete_cb` and `cb_arg` (since we cannot modify the in-memory layout of `RBD::AioCompletion`).

**Comment by trociny** [comment]
@dillaman need to update valgrind.supp. See [1]
```
2019-06-26T18:01:56.228 INFO:tasks.workunit.client.0.smithi145.stdout:[ RUN      ] TestLibRBD.TestIO
2019-06-26T18:01:56.232 INFO:tasks.workunit.client.0.smithi145.stdout:using new format!
2019-06-26T18:01:56.494 INFO:tasks.workunit.client.0.smithi145.stderr:==13293== Thread 28 fn-radosclient:
2019-06-26T18:01:56.494 INFO:tasks.workunit.client.0.smithi145.stderr:==13293== Conditional jump or move depends on uninitialised value(s)
2019-06-26T18:01:56.494 INFO:tasks.workunit.client.0.smithi145.stderr:==13293==    at 0x6F1E38: compare_exchange_weak (atomic:288)
2019-06-26T18:01:56.494 INFO:tasks.workunit.client.0.smithi145.stderr:==13293==    by 0x6F1E38: compare_exchange_weak (atomic:308)
2019-06-26T18:01:56.495 INFO:tasks.workunit.client.0.smithi145.stderr:==13293==    by 0x6F1E38: do_push<false> (queue.hpp:309)
2019-06-26T18:01:56.495 INFO:tasks.workunit.client.0.smithi145.stderr:==13293==    by 0x6F1E38: push (queue.hpp:271)
2019-06-26T18:01:56.495 INFO:tasks.workunit.client.0.smithi145.stderr:==13293==    by 0x6F1E38: librbd::io::AioCompletion::complete_external_callback() (AioCompletion.cc:252)
2019-06-26T18:01:56.495 INFO:tasks.workunit.client.0.smithi145.stderr:==13293==    by 0x6F33F7: librbd::io::AioCompletion::complete() (AioCompletion.cc:103)
2019-06-26T18:01:56.495 INFO:tasks.workunit.client.0.smithi145.stderr:==13293==    by 0x6F3ABF: librbd::io::AioCompletion::complete_request(long) (AioCompletion.cc:230)
2019-06-26T18:01:56.495 INFO:tasks.workunit.client.0.smithi145.stderr:==13293==    by 0x4EB0E8: Context::complete(int) (Context.h:77)
```
[1] http://pulpito.ceph.com/trociny-2019-06-26_16:27:34-rbd-wip-mgolub-testing-distro-basic-smithi/

**Comment by dillaman** [comment]
Tweaked the existing rule to cover both cases.

**Comment by trociny** [comment]
@dillaman Uh, there is still a valgrind warning for python test_aio_read [1]:
```
2019-06-28T07:48:07.613 INFO:tasks.workunit.client.0.smithi148.stderr:test_rbd.TestImage.test_aio_read ... ==12166== Thread 16 tp_librbd:
2019-06-28T07:48:07.613 INFO:tasks.workunit.client.0.smithi148.stderr:==12166== Conditional jump or move depends on uninitialised value(s)
2019-06-28T07:48:07.614 INFO:tasks.workunit.client.0.smithi148.stderr:==12166==    at 0x223EEF28: librbd::io::AioCompletion::complete_external_callback() (atomic:288)
2019-06-28T07:48:07.614 INFO:tasks.workunit.client.0.smithi148.stderr:==12166==    by 0x223F04E7: librbd::io::AioCompletion::complete() (AioCompletion.cc:103)
2019-06-28T07:48:07.614 INFO:tasks.workunit.client.0.smithi148.stderr:==12166==    by 0x223F0BAF: librbd::io::AioCompletion::complete_request(long) (AioCompletion.cc:230)
2019-06-28T07:48:07.614 INFO:tasks.workunit.client.0.smithi148.stderr:==12166==    by 0x222A3F58: Context::complete(int) (Context.h:77)
2019-06-28T07:48:07.614 INFO:tasks.workunit.client.0.smithi148.stderr:==12166==    by 0x15990CAE: ThreadPool::worker(ThreadPool::WorkThread*) (WorkQueue.cc:118)
2019-06-28T07:48:07.614 INFO:tasks.workunit.client.0.smithi148.stderr:==12166==    by 0x15991ACF: ThreadPool::WorkThread::entry() (WorkQueue.h:465)
2019-06-28T07:48:07.614 INFO:tasks.workunit.client.0.smithi148.stderr:==12166==    by 0x5209DD4: start_thread (in /usr/lib64/libpthread-2.17.so)
2019-06-28T07:48:07.615 INFO:tasks.workunit.client.0.smithi148.stderr:==12166==    by 0x5C24EAC: clone (in /usr/lib64/libc-2.17.so)
2019-06-28T07:48:07.615 INFO:tasks.workunit.client.0.smithi148.stderr:==12166==
2019-06-28T07:48:08.002 INFO:tasks.workunit.client.0.smithi148.stderr:ok
```
[1] http://pulpito.ceph.com/trociny-2019-06-28_06:54:35-rbd-wip-mgolub-testing-distro-basic-smithi/

**Comment by dillaman** [comment]
I suspect it's the ```pop``` operation, but it's odd that it only occurs for that one specific test case since all IO completions would follow the same code path.

**Comment by dillaman** [comment]
Turns out it was still the "push" operator, so I am going to try to fix this in a different way now by patching boost to fix the original bug. I have a branch building and will re-run the valgrind tests.

**Comment by dillaman** [comment]
http://pulpito.ceph.com/jdillaman-2019-06-28_17:02:52-rbd-wip-jd-testing-distro-basic-smithi/

**Pull request merged**